### ISSUE TITLE
Create "last n board states" feature as used in AlphaGo Zero

### ISSUE THREAD

**Issue by wrongu** [Rochester-NRT/RocAlphaGo#224]
This feature will contain `2*n` plains, with even planes corresponding to the current player's stones and odd planes corresponding to the other player's stones. Planes 0 and 1 will contain the current board state, 2 and 3 the previous state, 4 and 5 from 2 turns ago, etc..

On turn `t`, we should avoid any "undo" operations to reconstruct what the board looked like at turn `t-n`. The right way to do this is to simply shift the feature planes back by 2 each turn:

    current_feature[2:] = previous_feature[0:-2]
    current_feature[:2] = get_board_state[:2]

This is only tricky because our current feature preprocessing is not "stateful," so we do not have access to `previous_feature`. I propose the following design:

* The `Preprocessing` class will maintain a copy of the previous features that will be passed into the feature-generating functions.
* Add a flag to `Preprocessing.state_to_tensor` that, if `True`, indicates that the `state` argument has advanced by exactly 2 turns since the last time `state_to_tensor` was called. (Or, we may do this 'transparently' from inside `state_to_tensor` by looking at state history and hashes)
* Implement a fast "push-back" version of this feature, as described above, when possible. Else, fall back onto *very slow* reconstruction of the last `n` states from the game history.

**Comment by wrongu** [comment]
Depends on #220

**Comment by MaMiFreak** [comment]
we have two use cases:
- during MCTS ( we can create parent in GameState and mcts_do_move function: copy state, set parent and do move. that way we can use preprocessing as it is, this means additions GameState copies have to be created during game-converter usage )
- non MCTS like game-converter ( we could implement some sort of history in preprocessing, this means we need to copy preprocessor for each branch in MCTS )

To me MCTS case is most important, python is already slow we do not need anything slowing MCTS down any further. 
game-converter got a huge speedup because of the cython implementation. even with the gamestate copies being made it will be faster than the python implementation and there is even room for improvement with a multiprocessing implementation

**Comment by MaMiFreak** [comment]
> The right way to do this is to simply shift the feature planes back by 2 each turn

in addition to the shift every 2 planes also have to switch order ( current player, opponent player switches each turn )

**Comment by wrongu** [comment]
I agree MCTS speed is the priority, but copying the state after each move would be more expensive than creating a new `Preprocess` object per MCTS chain, no?

> in addition to the shift every 2 planes also have to switch order ( current player, opponent player switches each turn )

Similarly, we have separate `Preprocess` objects per NN (or, per player), so a given object would always create features from the perspective of the same player.

**Comment by MaMiFreak** [comment]
MCTS needs a copy of gamestate for each move expansion

**Comment by wrongu** [comment]
>MCTS needs a copy of gamestate for each move expansion

Why is that? It needs separate copies for each branch, but within a branch moves are played out linearly according to the current action probability `p(a|s)` at node `s`.

**Comment by MaMiFreak** [comment]
My interpretation of the paper is that their MCTS looks like

Expand tree n times:
- select most urgent move to explore (somewhere in tree)
- play move and preprocess features ( needs to copy gamestate before move is played )
- run prediction on preprocessed features
- expand tree with prediction data and gamestate

play move -> most visited child

no rollout is used

**Comment by wrongu** [comment]
If each node in the tree is a state, and each node stores `{action: N}, {action: Q}, {action: W}, {action: P}` maps, then the tree search can work like the following pseudo-code (ignore my poor choice of function names for now):

```python
init_policy_value = model.evaluate(root_state)
root_node = MCTSNode(init_policy_value)
for thread in range(1600):
    state = root_state.copy()  # The only place states need to be copied
    node = root_node
    for t in range(L):
        action = node.select()  # PUCT algorithm, updates virtual loss, etc...
        state.do_move(action)
        if not node.has_child(action):
            next_policy_value = model.evaluate(state)
            node.init_child(state, next_policy_value)  # Create a new node for the state resulting from 'action'
        node = node.children[action]
    # ... evaluate leaf node on a queue
    # ... backup result through the tree
# ... return (pi, value) of root node
```

I only see a need for `state.copy()` once per thread, and a linear progression of actions within each thread. Did I misunderstand your point?

**Comment by MaMiFreak** [comment]
i don't understand why you want to copy the rootstate..
you have to replay all the moves up to the expansion node..

also not sure why you have `for t in range(L):` what is this L?

please have a look at my selfplaytrainer code, right now my branch is not working but i have a working version with the same mcts code.
